### ISSUE TITLE
Multiplayer Stability Investigation - Have multiplayer stability issues been fixed since release 2019-07-17?

### ISSUE THREAD

**Issue by lfricken** [LoneGazebo/Community-Patch-DLL#8505]
I know VP has not typically been stable for multiplayer, but beyond normal desyncs, the first game I tried with a friend has completely broken and is now unplayable (Release 2019-07-17). When turn 81 loads neither of us can take any action (we can click UI, but nothing updates, units won't move, cities won't change production, etc.) It sits like this indefinitely. We tried backing up multiple turns, but on this turn it always fails.

I debugged it but wasn't exactly sure what to look for. I see the game update event getting fired, and even lua events triggering.

We went ahead and updated to the latest version, but before we commit multiple hours to a game: **Have these kinds of game freezing issues in multiplayer been addressed? Is the issue I'm describing known about?** Thank you!

P.S. I don't expect anyone to follow these reproduction steps, because we have our own mod, but:
* Remove any modpacks you have in **Sid Meier's Civilization V\Assets\DLC**
* Check out [this repo and branch](https://github.com/lfricken/LeonMod/tree/vp-2019-07-17-1.1) **INTO** your Civ5 Assets folder (notice the repo folder structure, and BE CAREFUL not to blindly have git delete existing folders, instead, clone the repo and then move it and merge directories)
* Extract and put the save file [Dido_0081 AD-1175-June.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/7799579/Dido_0081.AD-1175-June.zip) in **Documents\My Games\Sid Meier's Civilization 5\Saves\multi**
* Have another real human join your game as the 2nd player before launching, otherwise you won't see any issue

Follow up question, do you have advice (what to look at code wise) on fixing any of these desync/freeze issues? Is it feasible to fix them? I'd like to try fixing them, but I don't see where it's comparing deltas that cause desyncs, let alone freeze issues.

**Comment by ilteroi** [comment]
there have been lots of changes since 2019 ... that was ages ago. that said, multiplayer is not well tested. just give it a try!

**Comment by lfricken** [comment]
We just tried it and desyncs are worse. We have only hard crashed once, so we'll see whether we're able to finish the game.

I'd like to take a crack at fixing some of these multiplayer issues. Have people tried before? Should I be looking in the function [calculateSyncChecksum](https://github.com/LoneGazebo/Community-Patch-DLL/blame/master/CvGameCoreDLL_Expansion2/CvGame.cpp), or maybe debug/log calls to [CvDllNetMessageHandler.cpp](https://github.com/LoneGazebo/Community-Patch-DLL/blob/master/CvGameCoreDLL_Expansion2/CvDllNetMessageHandler.cpp)?

The strange thing about calculateSyncChecksum is that it appears to only get called via lua. Does anyone understand how the desync is ultimately compared across computers? Is that in lua somewhere, or in another dll?

**Comment by ilteroi** [comment]
honestly i have no idea how it is supposed to work. it seems events are sent to all players, so their gamestate should evolve in parallel. and the sync archives are only used to detect deviations ... that is all i can tell you, the root of the problem is that the network code is not part of the game core library and the source code is not public. so it's very frustrating. maybe the protocol can be reverse engineered by listening on the network but i've never tried.

**Comment by Cyclopia31** [comment]
Hi, I have tested some multiplayer games for the last few version and the game is actually unplayable. 
For 100 turns you can expect 50 Desyncs, and some CTDs for the player which isn't the Host. Start turn 100 (can be 150 for example) some AI start to hang during their turn (it can be a Civ AI or a city state AI). If you want a way to trigger a Desync you can play with Antic ruins on and take one of them with the "non-host" player. It will desync and give you another bonus (for example: You gain Mining tech with a ruin, the game desync and gives you 100 golds in the next turn instead of the tech).

I have an old VP (4.30) from 2018 and I can play it with 0 problem (desync etc) for 150 turns (I stopped the game turn 150, because it was just a "stability test" multiplayer game) so a stable experience with VP is possible ! 
Here is a link of that old modpack if you want to compare its behavior alongside a new modpack:[https://drive.google.com/file/d/1rWyInVqqKOS7-r65PuwhbnC1x3DvRTJ4/view?usp=sharing](url)

IanE9 had a suggestion where you can look to prevent desyncs : [https://github.com/LoneGazebo/Community-Patch-DLL/issues/8345](url)

If you find a way to improve the stability of VP in multiplayer I am sure that a lot of VP players will thank you !

Good luck !

**Comment by lfricken** [comment]
Thanks for all the info!

# Desyncs:
I did my own tests yesterday, both tests with No Randomness and found that current version had 0 desyncs over my 150 turn test if:
* No Barbarians
* No AI Civs

But did have desyncs immediately with them.

I did manage to recreate a separate issue though, in a game with NO barbs or ai civs, and no randomness. I have seen it with them as well:
# UI Freeze
In multiplayer only, sometimes the UI for both players will lock up after an AI round. Clicking buttons normally triggers code in [CvDllNetMessageHandler.cpp](https://github.com/LoneGazebo/Community-Patch-DLL/blob/master/CvGameCoreDLL_Expansion2/CvDllNetMessageHandler.cpp), but when this happens, no UI calls make it to the net message handler, so nothing in game happens. You can still get to some menus. If either player disconnects, the issue goes away for the remaining player, and they can advance the game state. [I have attached a save file that reliably reproduces it right after loading.](https://github.com/LoneGazebo/Community-Patch-DLL/files/7811299/Ahmad.al-Mansur_0123.AD-1260.zip)

Because it works even when loading the save, I suspect there is a boolean flag somewhere indicating that no players should be allowed to take actions (probably as a networking sanity/safety measure) but it's getting stuck _on_. Also, you can also see this behavior work correctly if you try to move units/do stuff during the AI turn, you get blocked, but only temporarily. Any idea where that logic is?
[Ahmad al-Mansur_0123 AD-1260.zip

**Comment by ilteroi** [comment]
there is a check for the acitve player in CvUnit::PushMission ... maybe that is a problem?

**Comment by lfricken** [comment]
> I have an old VP (4.30) from 2018...
Just tried it and we are having desyncs early, and every 4ish turns.

> there is a check for the acitve player in CvUnit::PushMission
I'll check that. Debugging the variable sync looked to be very difficult when I tried it.

**Comment by Cyclopia31** [comment]
> > I have an old VP (4.30) from 2018...
> 
> Just tried it and we are having desyncs early, and every 4ish turn
> 
I used it in local network, and had no problem. 
These desyncs problem are hard top understand :/

**Comment by Adonis-VII** [comment]
I've was playing been playing Civ V multiplayer every week with my mates since at least 2016 (first issue on Github at least).  The class of desyncs was originally related to things like scrolling the map too fast and unsynchronized rand() around barbarians and ruins. When these were addressed you'd expect to be able to play a game with a desync every 100-150 turns with stable Vox Populi for quite a while.  

However we have not played much since maybe before Covid due to what I believe to be a new class of desyncs. Around the time that ilteroi or Gazebo did a big unit movement overhaul new desyncs where when re-synced units would be in a different places from where non-host thought they should be. The game hasn't been stable for us since that set of changes.  I remember reading in one of the commits that the seed for the random functions is incorporates unit states?

**Comment by ilteroi** [comment]
hmm, i don't know about a "unit movement overhaul" ... there were lots of pathfinder improvements, but there is no random element there. 

the main thing we did was to replace many RNG calls which depend on the explicit pseudorandom seed with "fake RNG calls" which essentially use some piece of game state (eg hash of amount of gold plus the game turn) as seed to generate a random-looking number.

if there was a regression, it must have been something else. anyway i've come to believe that this whole RNG thing may be a red herring; should be easy to verify by experimentally returning 0 from all RNG calls (fake or not). but that doesn't help a lot in finding the real root cause.

**Comment by WardBenjamin** [comment]
I can confirm both the desync and UI freeze issues in multiplayer. Happy to help test any changes.

**Comment by Cyclopia31** [comment]
> Thanks for all the info!
> 
> # Desyncs:
> I did my own tests yesterday, both tests with No Randomness and found that current version had 0 desyncs over my 150 turn test if:
> 
> * No Barbarians
> * No AI Civs
> 
> But did have desyncs immediately with them, so the core issue must be somewhere in there. I'll try debugging the desync code you linked.
> 
> I did manage to recreate a separate issue though, in a game with NO barbs or ai civs, and no randomness. I have seen it with them as well:
> 
> # UI Freeze
> In multiplayer only, sometimes the UI for both players will lock up after an AI round. Clicking buttons normally triggers code in [CvDllNetMessageHandler.cpp](https://github.com/LoneGazebo/Community-Patch-DLL/blob/master/CvGameCoreDLL_Expansion2/CvDllNetMessageHandler.cpp), but when this happens, no UI calls make it to the net message handler, so nothing in game happens. You can still get to some menus. If either player disconnects, the issue goes away for the remaining player, and they can advance the game state. [I have attached a save file that reliably reproduces it right after loading.](https://github.com/LoneGazebo/Community-Patch-DLL/files/7811299/Ahmad.al-Mansur_0123.AD-1260.zip)
> 
> Because it works even when loading the save, I suspect there is a boolean flag somewhere indicating that no players should be allowed to take actions (probably as a networking sanity/safety measure) but it's getting stuck _on_. Also, you can also see this behavior work correctly if you try to move units/do stuff during the AI turn, you get blocked, but only temporarily. Any idea where that logic is?

Hey, did you find anything ? :)
